compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -300,60 +301,85 @@ fn negative_impl<'cx, 'tcx>(
-            &infcx,
-            FulfillmentContext::new(),
-            ObligationCause::dummy(),
-            impl1_env,
-            impl1_trait_ref,
-        ) {
-            Ok(impl1_trait_ref) => impl1_trait_ref,
-            Err(err) => {
-                bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+        // create a parameter environment corresponding to a (placeholder) instantiation of impl1
+        let impl_env = tcx.param_env(impl1_def_id);
+        let subject1 = match tcx.impl_subject(impl1_def_id) {
+            ImplSubject::Trait(impl1_trait_ref) => {
+                match traits::fully_normalize(
+                    &infcx,
+                    FulfillmentContext::new(),
+                    ObligationCause::dummy(),
+                    impl_env,
+                    impl1_trait_ref,
+                ) {
+                    Ok(impl1_trait_ref) => ImplSubject::Trait(impl1_trait_ref),
+                    Err(err) => {
+                        bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+                    }
+                }
             }
+            subject @ ImplSubject::Inherent(_) => subject,
         };
 
+        let (subject2, obligations) =
+            impl_subject_and_obligations(&infcx, impl_env, subject1, impl2_def_id);
+
+        !equate(&infcx, impl_env, impl1_def_id, subject1, subject2, obligations)
+    })
+}
+
+fn impl_subject_and_obligations<'cx, 'tcx>(
+    infcx: &InferCtxt<'cx, 'tcx>,
+    impl_env: ty::ParamEnv<'tcx>,
+    subject1: ImplSubject<'tcx>,
+    impl2_def_id: DefId,
+) -> (ImplSubject<'tcx>, Box<dyn Iterator<Item = PredicateObligation<'tcx>> + 'tcx>) {
+    if let ImplSubject::Trait(_) = subject1 {
         // Attempt to prove that impl2 applies, given all of the above.
         let selcx = &mut SelectionContext::new(&infcx);
         let impl2_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl2_def_id);
         let (impl2_trait_ref, obligations) =
-            impl_trait_ref_and_oblig(selcx, impl1_env, impl2_def_id, impl2_substs);
-
-        // do the impls unify? If not, not disjoint.
-        let Ok(InferOk { obligations: more_obligations, .. }) = infcx
-            .at(&ObligationCause::dummy(), impl1_env)
-            .eq(impl1_trait_ref, impl2_trait_ref)
-        else {
-            debug!(
-                "explicit_disjoint: {:?} does not unify with {:?}",
-                impl1_trait_ref, impl2_trait_ref
-            );
-            return false;
-        };
+            impl_trait_ref_and_oblig(selcx, impl_env, impl2_def_id, impl2_substs);
 
-        let opt_failing_obligation = obligations
-            .into_iter()
-            .chain(more_obligations)
-            .find(|o| negative_impl_exists(selcx, impl1_env, impl1_def_id, o));
+        (ImplSubject::Trait(impl2_trait_ref), Box::new(obligations))
+    } else {
+        (infcx.tcx.impl_subject(impl2_def_id), Box::new(iter::empty()))
```

Comment:
hmm, I *think* it can have obligations, e.g.

```rust
impl<T> Vec<T>
where
    T: Ord,
```

would need `T: Ord`.

Reply:
Aren't those filled in here https://github.com/rust-lang/rust/pull/95082/files#diff-8d9bb4ed968c5734cb6004b14afb6e7652d192f6db4a5fa9758c7d9abc05d7f8R371 ?

Reply:
Actually, no, those are different. Those obligations are obligations that arise as a side-effect of equating two types, not the where clauses from the impl. But before you make any change here we should be able to create a test.

Reply:
Fixed.